Add Component for PlayStation 4 consoles as integration/config flow

ktnrg45 opened the pull request:
## Description:
Configures via config flow/integrations.
Add component media_player.ps4. Reopened new PR as commits were cluttered. @ #19469
This component grants control and state updates for a Sony PlayStation 4 console.
It masquerades as the official PS4 Second Screen App which is only possible after capturing the app's credential token and using a pin code from the PS4 to become an authorized device.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8146

## Example entry for `configuration.yaml` (if applicable):
Does not support entries from `configuration.yaml`

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

kennedyshead commented:
It cant find media for region R2 CUSA00265 minecraft

ktnrg45 commented:
> It cant find media for region R2 CUSA00265 minecraft

As I do not have a PS4 with the same region. I will need some more information from your entity states.
Under the entity states tab in your instance, find the entity for your media_player.playstation_4 and please post the following states when running your game:
media_content_id: " "
media_title: " "

Also do any other games work or is it just that one in particular?

ktnrg45 commented:
> Same question as on #19469 - want to try this again but can't get config dialog popping up. How do I go about resetting it so it will ?

Delete .ps4.conf from your /config directory. Restart homeassistant. In your frontend there is a bell/alarm icon at the top right. This displays your notifications in Lovelace UI. Click "configure" on the entry "PS4 Setup". Make sure you press the "Start" Button in the popup window.

FYI, verified again that setup is working right now on my Hass.OS instance.

UnrealKazu commented:
Great work! I've been eager to integrate my PS4 into HASS for some time.

I've been testing this component using my Docker-driven HASS setup. There are still some difficulties if you're running HASS using Docker. I guess the `NET_BIND_SERVICE` requirement makes things difficult, because I assume the HASS container needs to run in privileged mode for this to work. Whenever I start the configuration, an event is fired, but the instance does not pop up in my Second Screen app. What is the easiest way to debug this? Can I somehow check if the python script is able to connect to the socket?

An actual bug I've found is when you click on `configure` in the PS4 card, click on `start` so it starts its listener, but let it run out (in my case because of the Docker difficulties, but that's irrelevant here). If you then close the configuration popup, re-open it and click on start again, you get an error. The stack trace logged by HASS is as follows:

```
Error handling message: {'type': 'call_service', 'domain': 'configurator', 'service': 'configure', 'service_data': {'configure_id': '140458694581888-1', 'fields': {}}, 'id': 18}

Traceback (most recent call last):
  File "/usr/src/app/homeassistant/components/websocket_api/decorators.py", line 17, in _handle_async_response
    await func(hass, connection, msg)
  File "/usr/src/app/homeassistant/components/websocket_api/commands.py", line 148, in handle_call_service
    connection.context(msg))
  File "/usr/src/app/homeassistant/core.py", line 1121, in async_call
    self._execute_service(handler, service_call))
  File "/usr/src/app/homeassistant/core.py", line 1143, in _execute_service
    await handler.func(service_call)
  File "/usr/src/app/homeassistant/components/configurator.py", line 221, in async_handle_service_call
    call.data.get(ATTR_FIELDS, {}))
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/media_player/ps4.py", line 81, in configuration_callback
    result = credentials.listen()
  File "/config/deps/lib/python3.6/site-packages/pyps4_homeassistant/ps4.py", line 332, in listen
    data = response[0]
UnboundLocalError: local variable 'response' referenced before assignment```

ktnrg45 commented:
@UnrealKazu 
Thanks for your feedback. I'm not very familiar with docker so I'm not sure if I can help. Running your container as privileged is a solution although I'm not sure if there are security issues in doing so, so I cannot recommend doing so. Basically you have to be able to bind to ports 987 and 997. To test if your container as access. Run the python runtime/cli inside of your container. Then type the following:
`import socket`
`sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)`
`sock.bind(('0.0.0.0', 987))`
If you get no errors your container has access to that port and all should be good.
If your container does not have access you will probably get an error like:
`Traceback (most recent call last):`
`  File "<stdin>", line 1, in <module>`
`  File "/usr/lib/python2.7/socket.py", line 228, in meth`
`    return getattr(self._sock,name)(*args)`
`socket.error: [Errno 13] Permission denied`

To answer your other question, if the configurator reaches the timeout as you have encountered, you have to restart your home assistant instance and it should work again. I'll see if I can fix this.

UnrealKazu commented:
@ktnrg45 Thanks for your comment. Based on your information, I was able to verify that inside the container, a successful socket connection could be made. So what I did next was open up both the ports you mentioned (987 and 997) for UDP and TCP on the HASS container itself. The UDP part is crucial because this is the port which the Second Screen app monitors on.

Now Home-Assistant does appear in the Second Screen app. Once I click on it in the app, the configuration pop-up disappears in HASS and I can click on `Pair PS4`. However, the Second Screen app keeps 'Connecting' to the HASS instance. I assume this is not the normal behaviour, or is the pairing part necessary to complete the connecting phase?

I'm getting a `Socket error` warning once I enter the PS4's pin in the `Pair PS4` step, but did not have the time to look into it further. I will do so in the next couple of days, just wanted to keep you updated on my progress of testing it in a Dockerized environment.

ktnrg45 commented:
> @ktnrg45 Thanks for your comment. Based on your information, I was able to verify that inside the container, a successful socket connection could be made. So what I did next was open up both the ports you mentioned (987 and 997) for UDP and TCP on the HASS container itself. The UDP part is crucial because this is the port which the Second Screen app monitors on.
> 
> Now Home-Assistant does appear in the Second Screen app. Once I click on it in the app, the configuration pop-up disappears in HASS and I can click on `Pair PS4`. However, the Second Screen app keeps 'Connecting' to the HASS instance. I assume this is not the normal behaviour, or is the pairing part necessary to complete the connecting phase?
> 
> I'm getting a `Socket error` warning once I enter the PS4's pin in the `Pair PS4` step, but did not have the time to look into it further. I will do so in the next couple of days, just wanted to keep you updated on my progress of testing it in a Dockerized environment.

@UnrealKazu The Second Screen app should not complete the connection as it is unnecessary as the credentials should have already be sent to Home Assistant. This is normal. As far as the socket.error exception goes I'm not sure without more info. Make sure your ps4 ip is correct in your configuration.yaml and that you can bind to port 997 as TCP.

glassbase commented:
I have been using the other dev’s component that you forked this from without issue. I deleted my .ps4.conf and switched to trying yours out. 

I receive this in my log and does not setup. I have removed .ps4.conf and redid setup a few times to see if that’s the issue. 

2019-01-20 08:07:52 ERROR (MainThread) [homeassistant.components.media_player] ps4: Error on device update!
Traceback (most recent call last):
  File "/usr/src/app/homeassistant/helpers/entity_platform.py", line 248, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/usr/src/app/homeassistant/helpers/entity.py", line 349, in async_device_update
    await self.hass.async_add_executor_job(self.update)
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/app/homeassistant/util/__init__.py", line 315, in wrapper
    result = method(*args, **kwargs)
  File "/usr/src/app/homeassistant/util/__init__.py", line 315, in wrapper
    result = method(*args, **kwargs)
  File "/config/custom_components/media_player/ps4.py", line 186, in update
    self.get_title_data(titleid, name)
  File "/config/custom_components/media_player/ps4.py", line 228, in get_title_data
    name, titleid, self._region)
  File "/config/deps/lib/python3.6/site-packages/pyps4_homeassistant/ps4.py", line 262, in get_ps_store_data
    full_id = full_id[1]
IndexError: list index out of range

ktnrg45 commented:
@glassbase Thanks for your comment. The error refers to fetching the correct artwork for a game from Sonys database. Try closing all games on your PS4 and restart home assistant and see if there are any exceptions raised to see if the issue is caused elsewhere. If there are no errors than the issue is retrieving the artwork. Make sure in your configuration.yaml you have included your correct 'region' See referenced docs.

glassbase commented:
@ktbrg45 With PS4 off, the component initialized. I was able to turn on off the PS4 and see the title of the game being played. 

No image art with same error
  File "/config/deps/lib/python3.6/site-packages/pyps4_homeassistant/ps4.py", line 262, in get_ps_store_data
    full_id = full_id[1]
IndexError: list index out of range

Game is Black Ops 4, CUSA11100

ktnrg45 commented:
> @ktbrg45 With PS4 off, the component initialized. I was able to turn on off the PS4 and see the title of the game being played.
> 
> No image art with same error
> 
> ```
>   File "/config/deps/lib/python3.6/site-packages/pyps4_homeassistant/ps4.py", line 262, in get_ps_store_data
>     full_id = full_id[1]
> IndexError: list index out of range
> ```
> Game is Black Ops 4, CUSA11100

@glassbase It is a bug querying that particular title. I will add a fix so that it wont stop the component from running.

ktnrg45 commented:
> > @ktbrg45 With PS4 off, the component initialized. I was able to turn on off the PS4 and see the title of the game being played.
> > No image art with same error
> > ```
> >   File "/config/deps/lib/python3.6/site-packages/pyps4_homeassistant/ps4.py", line 262, in get_ps_store_data
> >     full_id = full_id[1]
> > IndexError: list index out of range
> > ```
> > Game is Black Ops 4, CUSA11100
> 
> @glassbase It is a bug querying that particular title. I will add a fix so that it wont stop the component from running.

In the meantime if you have access to your python runtime you could try doing this.
Run  the python cli and run cmds:
```
import pyps4_homeassistant as pyps4
ps4 = pyps4.Ps4("Your PS4 IP Address", "Your Credentials in .ps4.conf (the long string)")
ps4.get_status()

```
Then post the output for the following items (please dont post the other items for your security):
```
running-app-name
running-app-titleid

```
Should look somewhat like this:
`'running-app-name': 'Call of Duty®', 'running-app-titleid': 'CUSA11100'`

If you can copy/paste and post the exact output you get for that info I can further troubleshoot the issue.

glassbase commented:
Yes that's what it is. ` 'running-app-name': 'Call of Duty®', 'running-app-titleid': 'CUSA11100'`

ktnrg45 commented:
> Yes that's what it is. ` 'running-app-name': 'Call of Duty®', 'running-app-titleid': 'CUSA11100'`
> 
> Full output if you need it:
> `{'status_code': 200, 'status': 'Ok', 'host-id': 'BC60A7E15CBB', 'host-type': 'PS4', 'host-name': 'PS4', 'host-request-port': '997', 'device-discovery-protocol-version': '00020020', 'system-version': '06200011', 'running-app-name': 'Call of Duty®', 'running-app-titleid': 'CUSA11100', 'host-ip': '10.0.0.7'}`
> 
> I am no expert but can read a little code. Searching using that valkyrie-api url with "[call of duty](https://store.playstation.com/valkyrie-api/en/us/19/faceted-search/call%20of%20duty)" does not return CUSA11000 in results. I was thinking it was that the registered symbol on name but it is not.

Yep. Exactly.  For some reason this game needs the full title inputted to show up in the search. Theres a bug in the library which creates incomplete titles.  Ill have a fix out today.

ktnrg45 commented:
> Yes that's what it is. ` 'running-app-name': 'Call of Duty®', 'running-app-titleid': 'CUSA11100'`
> 
> Full output if you need it:
> `{'status_code': 200, 'status': 'Ok', 'host-id': 'BC60A7E15CBB', 'host-type': 'PS4', 'host-name': 'PS4', 'host-request-port': '997', 'device-discovery-protocol-version': '00020020', 'system-version': '06200011', 'running-app-name': 'Call of Duty®', 'running-app-titleid': 'CUSA11100', 'host-ip': '10.0.0.7'}`
> 
> I am no expert but can read a little code. Searching using that valkyrie-api url with "[call of duty](https://store.playstation.com/valkyrie-api/en/us/19/faceted-search/call%20of%20duty)" does not return CUSA11000 in results. I was thinking it was that the registered symbol on name but it is not.

@glassbase Applied fix. Should work now. I got correct artwork for COD: WWII.

kennedyshead commented:
@ktnrg45 Happy times! The fix worked for R2

kennedyshead commented:
On/Off stopped working though 
```
Traceback (most recent call last):
  File "/home/***/src/home-assistant/env/lib/python3.7/site-packages/homeassistant-0.86.0.dev0-py3.7.egg/homeassistant/helpers/service.py", line 287, in _handle_service_platform_call
    await getattr(entity, func)(**data)
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/***/.homeassistant/custom_components/media_player/ps4.py", line 346, in turn_off
    self._ps4.standby()
  File "/home/***/src/home-assistant/env/lib/python3.7/site-packages/pyps4_homeassistant/ps4.py", line 92, in standby
    self._connection.standby()
  File "/home/***/src/home-assistant/env/lib/python3.7/site-packages/pyps4_homeassistant/connection.py", line 76, in standby
    msg = self._decipher.decrypt(msg)
AttributeError: 'NoneType' object has no attribute 'decrypt'
```

glassbase commented:
All working with COD Black Ops 4 now (not WWII ;) )

Kvasenok commented:
Man, I hoped it would be implemented in 0.86 release. Oh well, guess we'll have to wait a bit more))

kennedyshead commented:
> Man, I hoped it would be implemented in 0.86 release. Oh well, guess we'll have to wait a bit more))

Use it as a custom component for before release ;)

Kvasenok commented:
@kennedyshead I already just. It just doesn't seem to be very stable. For some reason it won't let me control my ps4 from time to time, but sometimes it works ok. Although I haven updated the custom a long time ago, I just think the official component should be more stable in any way)

kennedyshead commented:
> @kennedyshead I already do. It just doesn't seem to be very stable. For some reason it won't let me control my ps4 from time to time, but sometimes it works ok. Although I have updated the custom component a long time ago, I just think the official component should be more stable in any way)

I meant this version: https://raw.githubusercontent.com/home-assistant/home-assistant/4232d34edebc87a208701db6c814e6157de9a737/homeassistant/components/media_player/ps4.py :)

It is fairly stable, I do ave some problems with control, as described above

falkyre commented:
Quick question, will this handle more than one PS4 on the network?

ktnrg45 commented:
> Quick question, will this handle more than one PS4 on the network?

@falkyre I haven't tested as I don't have more than one. Feel free to try. Your credentials should stay the same if using the same PSN ID. Just move the .ps4.conf file and delete the entry from your configuration.yaml after you set up the first ps4, and then after you set up the second, merge the data from both files like:

```
{
    "IP of PS4 #1": "Your Credentials"
    "IP of PS4 #2": "Your Credentials"
}
```

Then re-add your entry for PS4 #1 to your configuration.yaml.

RutgerJoosten commented:
This will be awsome when it is intergrated!
But can I get this to work as a custom component without some manual SSH install? (Of yes, hoe?) Or do I have to wait for the component to be added in hass? (Running hassio atm)
Tnx!

falkyre commented:
> > Quick question, will this handle more than one PS4 on the network?
> 
> @falkyre I haven't tested as I don't have more than one. Feel free to try. Your credentials should stay the same if using the same PSN ID. Just move the .ps4.conf file and delete the entry from your configuration.yaml after you set up the first ps4, and then after you set up the second, merge the data from both files like:
> 
> ```
> {
>     "IP of PS4 #1": "Your Credentials"
>     "IP of PS4 #2": "Your Credentials"
> }
> ```
> Then re-add your entry for PS4 #1 to your configuration.yaml.

So I tried setting up with multiple PS4s.  Initial setup once the .ps4.conf file is in the proper format, both PS4s are controllable, I can turn them off and turn them on with no issues.  However, once one of te PS4s starts doing something, the activity is shown for both PS4s.  Even if I start an game or app on the second PS4, the two media_player entries show what's going on on the first PS4 to start doing something.  Also, the .ps4-games.json is only updated by the first PS4 to start doing something.  I have debug logging turned on for the custom_components.media_player.ps4 but nothing is showing in the logs at all.  I had a similar problem with the ps4-waker docker image for hassio.  It seems it only recognizes the first PS4 on the network that starts doing something.

ktnrg45 commented on homeassistant/components/media_player/ps4.py:
```diff
@@ -0,0 +1,369 @@
+
+def request_configuration(hass, config, add_devices, _creds):
+    """Request configuration steps from the user."""
+    configurator = hass.components.configurator
```
@MartinHjelmare Added config flow. How does it look?

UnrealKazu commented:
I got around to finally diving into the issues I was having. Docker users should have the HASS Docker run on their host network (`--net=host` [docker run command] or `network_mode: "host"` [docker-compose command]). As far as I know, this is the only way for HASS to reach the PS4 and connect with it. It took me some fiddling around with both this component and the separate `ps4-waker` CLI in different Docker containers to reach this conclusion. If you want, I can add a Docker-specific description to your documentation PR?

I've already tested your new integration-style version you've just committed. It is working nicely! I can turn the PS4 on and off, and it correctly tracks the playing state of the PS4. The new configuration flow is also great :) Awesome work!

ktnrg45 commented:
> I got around to finally diving into the issues I was having. Docker users should have the HASS Docker run on their host network (`--net=host` [docker run command] or `network_mode: "host"` [docker-compose command]). As far as I know, this is the only way for HASS to reach the PS4 and connect with it. It took me some fiddling around with both this component and the separate `ps4-waker` CLI in different Docker containers to reach this conclusion. If you want, I can add a Docker-specific description to your documentation PR?
> 
> I've already tested your new integration-style version you've just committed. It is working nicely! I can turn the PS4 on and off, and it correctly tracks the playing state of the PS4. The new configuration flow is also great :) Awesome work!

Awesome! Yeah that'd be great. Thanks for finding that out.

furiaceka87 commented:
I tried this plugin after last commit and it works like a charm, what modification is required to merge in 0.87?

ktnrg45 commented:
> I tried this plugin after last commit and it works like a charm, what modification is required to merge in 0.87?
@furiaceka87 Glad it works. Just waiting for component to be reviewed and approved @MartinHjelmare .

ktnrg45 commented:
> Hey @ktnrg45,
> 
> I feel like that logic to get game title and artwork needs to be tweaked. I am playing NHL 19 yet it says the 99 Edition. The 99 Edition is the first one listed in [valkyrie-api](https://store.playstation.com/valkyrie-api/en/us/19/faceted-search/nhl%2019) which I believe it why it is grabbed.
> 
> from states page:
> 
> ```
>   "media_content_id": "CUSA11126",
>   "media_title": "NHL® 19 99 Edition",
> ```
> I think you should use "game-content-type":"Full Game" in the valkyrie-api functions for better title and artwork identification, if that's possible. (99 Edition has "game-content-type":"Bundle")
> 
> Sorry if it seems I am nitpicking, just helping you work out the kinks. Cheers

@glassbase Updated pypi library. Try now and see if it works.

ktnrg45 commented:
@balloob @fabaff @frenck Can I get one of you guys to review my PR? I havent got much, if any input at all from you guys for a month. Id like to know if theres anything I can improve or change. Its been hard to iterate my code with long spans of no feedback, especially with all of the architecture changes. Thanks.

Kane610 commented on homeassistant/components/ps4/media_player.py:
```diff
@@ -0,0 +1,390 @@
+        _sw_version = _sw_version[1:4]
+        sw_version = _sw_version[0] + '.' + _sw_version[1:]
+        self._info = {
+            'config_entry_id': config.entry_id,
```
An entity doesn't provide config entry id. Look how it works for other entities

ktnrg45 replied:
Oops. Didn't know why I put that there.

Kane610 commented on .coveragerc:
```diff
@@ -374,6 +374,8 @@ omit =
     homeassistant/components/plum_lightpad/*
     homeassistant/components/point/*
+    homeassistant/components/prometheus.py
+    homeassistant/components/ps4/*    
```
Config entries require testing

Kane610 reviewed (changes requested):
I think you should simplify this PR. Only support one device, remove everything not tied specifically to functionality of device and entity and create tests.

ktnrg45 commented on .coveragerc:
```diff
@@ -374,6 +374,8 @@ omit =
     homeassistant/components/pilight/*
     homeassistant/components/plum_lightpad/*
     homeassistant/components/point/*
+    homeassistant/components/prometheus.py
```
Ok. Sounds good.

ktnrg45 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,215 @@
+REQUIREMENTS = ['pyps4-homeassistant==0.2.5']
+
+
+async def async_setup(hass, config):
```
Ok. So youre saying that the config flow will be called by the core right?

Kane610 replied:
Config flow is initiated by either importing from configuration.yaml, discovery or from the integrations page in the front end.

The import is mainly when you have an existing configuration and you are porting the component over to config entries.

You can of course keep this, but it is not really providing any benefit

ktnrg45 replied:
Ok. Thanks that makes sense.

MartinHjelmare commented on homeassistant/components/ps4/media_player.py:
```diff
@@ -0,0 +1,390 @@
+        if status is not None:
+            self._retry = 0
+            if status.get('status') == 'Ok':
+                titleid = status.get('running-app-titleid')
```
Please separate words with underscore in variable names. Let's call this `title_id`.

MartinHjelmare commented on homeassistant/components/ps4/media_player.py:
```diff
@@ -0,0 +1,390 @@
+        self._state = STATE_UNKNOWN
+        _LOGGER.debug("PS4 could not be reached")
+        self._retry = 0
+        self.update()
```
Why do we continue calling update after the retries have failed? Won't this interfere with the normal update interval?

ktnrg45 replied:
I figured if the PS4 dropped connection it would  be a faster update but I guess it would just throw a bunch of errors and overload the thread. I'll remove.

MartinHjelmare commented on homeassistant/components/ps4/media_player.py:
```diff
@@ -0,0 +1,390 @@
+        """Set states for state off."""
+        self.no_title()
+        self._state = STATE_OFF
+        self.update()
```
Why do we call update again here?

rytilahti commented on homeassistant/components/ps4/media_player.py:
```diff
@@ -0,0 +1,390 @@
+
+
+DEPENDENCIES = ['ps4']
+REQUIREMENTS = ['pyps4-homeassistant==0.2.5']
```
You can remove the REQUIREMENTS here as it is already done in the component, don't forget to update `requirements_all.txt` after removing it.

rytilahti commented on homeassistant/components/ps4/media_player.py:
```diff
@@ -0,0 +1,390 @@
+        self._retry = 0
+        self.update()
+
+    def no_title(self):
```
Would `reset_title` or even `reset_state` be a better name here?

ktnrg45 replied:
ok. changed.

rytilahti commented on .coveragerc:
```diff
@@ -451,6 +451,7 @@ omit =
     homeassistant/components/sensor/fritzbox_netmonitor.py
     homeassistant/components/sensor/gearbest.py
     homeassistant/components/sensor/geizhals.py
+    homeassistant/components/sensor/github.py
```
Remove unrelated changes, including this the xs1 and zha changes below.

ktnrg45 replied:
I'm not sure how to undo this. I don't need to have the file changed, so I copied the coveragerc file from the dev branch of my fork is this okay or is there a better way?

ktnrg45 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,215 @@
+
+from collections import OrderedDict
+import logging
+import voluptuous as vol
```
Ok. Removed.

ktnrg45 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,215 @@
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/ps4/
+"""
+
```
Removed.

ktnrg45 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,215 @@
+DEFAULT_NAME = "PlayStation 4"
+DEFAULT_REGION = "R1"
+
+
```
Removed.

ktnrg45 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,215 @@
+
+async def async_setup_entry(hass, config_entry):
+    """Set up PS4 from a config entry."""
+    if hass.config_entries.async_entries(DOMAIN):
```
Hmmm async_setup_entry would be called when new entry is created right? Ill remove that part.

Kane610 replied:
Yes, it is initiated by user from integrations page

ktnrg45 commented on homeassistant/components/ps4/media_player.py:
```diff
@@ -0,0 +1,390 @@
+Support for PlayStation 4 consoles.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/ps4/
```
Ok. Got confused because I was looking at Cast and Sonos and they didn't have the docs split into types.

ktnrg45 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,215 @@
+                )
+
+        # Show User Input form.
+        link_schema = OrderedDict()
```
Whats wrong with it exactly? I copied the format from the docs:
```
    async def async_step_user(self, user_input=None):
        # Use OrderedDict to guarantee order of the form shown to the user
        data_schema = OrderedDict()
        data_schema[vol.Required('username')] = str
        data_schema[vol.Required('password')] = str

        return self.async_show_form(
            step_id='init',
            data_schema=vol.Schema(data_schema)
        )
```

Heres the schema from Unifi:
```
        return self.async_show_form(
            step_id='user',
            data_schema=vol.Schema({
                vol.Required(CONF_HOST): str,
                vol.Required(CONF_USERNAME): str,
                vol.Required(CONF_PASSWORD): str,
                vol.Optional(CONF_PORT, default=DEFAULT_PORT): int,
                vol.Optional(
                    CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): bool,
            }),
            errors=errors,
        )
```
The only difference I can see is that Unifi defines the schema in the arguments for show_form. If I wanted my form to be ordered using OrderedDict from my understanding of it, I have to define the schema as OrderedDict and then append the dict keys+values in the intended order. How would this work for in an argument?

Kane610 replied:
Good enough I guess, I prefer to keep everything as similar as possible

ktnrg45 replied:
Ok sounds good.

ktnrg45 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,215 @@
+        except pyps4.errors.LoginFailed:
+            return False
+        except pyps4.errors.NotReady:
+            return 'not_ready'
```
I just returned two boolean vars. Hope thats okay.

MartinHjelmare replied:
It's ok. I suggested a namedtuple to allow adding more return values later without breaking code that relies on the method. But since this is a private method, I don't think we have to worry about that.

Kane610 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,206 @@
+
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_CODE, CONF_HOST, CONF_IP_ADDRESS,
```
Don't break row before you have to.

ktnrg45 commented on homeassistant/components/ps4/media_player.py:
```diff
@@ -0,0 +1,390 @@
+        self._media_title = None
+        self._media_image = None
+        self._source = None
+        self._games = self.load_games()
```
Defined as empty list and dict instead of function

ktnrg45 commented:
> I think you should simplify this PR. Only support one device, remove everything not tied specifically to functionality of device and entity and create tests.

Yeah I agree supporting multiple devices is adds a lot of complexity that most users probably won't use.

ktnrg45 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,215 @@
+async def async_setup_entry(hass, config_entry):
+    """Set up PS4 from a config entry."""
+    if hass.config_entries.async_entries(DOMAIN):
+        hass.data[DOMAIN] = config_entry
```
Removed. Was using for multiple device setup.

ktnrg45 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,215 @@
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class PlayStation4FlowHandler(config_entries.ConfigFlow):
```
Moved.

ktnrg45 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,215 @@
+        self.creds = await self.hass.async_add_executor_job(
+            credentials.listen)
+
+    async def _async_port_bind(self, port):
```
Moved.

ktnrg45 commented on homeassistant/components/ps4/media_player.py:
```diff
@@ -0,0 +1,390 @@
+def _setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up PS4 Platform."""
+    import pyps4_homeassistant as pyps4
+    hass.data[PLATFORM] = PS4Data()
```
Changed it to new key

ktnrg45 commented:
Thanks for the reviews. They were really helpful and  cleared some things up.

Im starting on the config flow tests. Is there any guidelines for these? I couldn't find any docs for them. @Kane610 @MartinHjelmare

RutgerJoosten commented:
@ktnrg45 I can't speak for anyone else, but support for multiple divices would be really apreaciated. We have 2 ps4s next to each other and it would be nice if they are both in hass(io). I can think of multiple automations that wel be very handy. For now it is good to focus on 1 device to get it al working, but I hope you will concider adding multiple divices setup.

falkyre commented:
I second the multiple devices as well.

On Mon., Feb. 11, 2019, 6:29 p.m. Rutger Joosten, <notifications@github.com>
wrote:

> @ktnrg45 <https://github.com/ktnrg45> I can't speak for anyone else, but
> support for multiple divices would be really apreaciated. We have 2 ps4s
> next to each other and it would be nice if they are both in hass(io). I can
> think of multiple automations that wel be very handy. For now it is good to
> focus on 1 device to get it al working, but I hope you will concider adding
> multiple divices setup.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/20007#issuecomment-462554465>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGsCFobqZUPe1xcGwG46EJMABzZ9otwRks5vMgr4gaJpZM4Z8itz>
> .
>

ktnrg45 commented:
> @ktnrg45 I can't speak for anyone else, but support for multiple divices would be really apreaciated. We have 2 ps4s next to each other and it would be nice if they are both in hass(io). I can think of multiple automations that wel be very handy. For now it is good to focus on 1 device to get it al working, but I hope you will concider adding multiple devices setup.

Yeah of course. I originally written the code to support multiple devices, but had a few issues. Right now its easier to omit the code and get the component out.

MartinHjelmare commented:
@ktnrg45 look eg at hue config flow tests.

ktnrg45 commented:
@MartinHjelmare @Kane610 Created test for config_flow.py

Kane610 commented on homeassistant/components/ps4/__init__.py:
```diff
@@ -0,0 +1,36 @@
+    """Set up the PS4 Component."""
+    _LOGGER.debug("Registering PlayStation 4 Config Flow Version %s",
+                  PlayStation4FlowHandler.VERSION)
+    hass.data[DOMAIN] = {}
```
No reason of doing this here, weird dependency since you don't really need this function at all.

ktnrg45 replied:
I know. When testing after I separated the config_flow, I get an exception for flow not found unless I imported the flow explicitly. Perhaps Ill retry after rebase.

MartinHjelmare replied:
We have to import the flow into the component to make sure the flow module is read and the flow registered. We can disable the pylint/flake8 warning about not used import on that line.

ktnrg45 replied:
Oh. Ok. Removed that line.

ktnrg45 replied:
Everything should be ready now.

MartinHjelmare replied:
This isn't used, so we can remove this line.

Kane610 commented on homeassistant/components/ps4/media_player.py:
```diff
@@ -0,0 +1,375 @@
+        if self._media_content_id is None:
+            return MEDIA_IMAGE_DEFAULT
+        try:
+            return self._media_image
```
Is this really a dict?

ktnrg45 replied:
Lol no. Was leftover code from an old version.

Kane610 reviewed (commented):
Much easier to review now 👍

ktnrg45 commented on .coveragerc:
```diff
@@ -10,32 +10,455 @@ omit =
     homeassistant/helpers/signal.py
 
     # omit pieces of code that rely on external devices being present
-    homeassistant/components/abode/*
```
Ok rebased my branch and corrected all conflicts. Can't tell if its proper still.

Kane610 replied:
Well there are still a lot of changes in coveragerc that has nothing to do with this PR. We don't want to get any historic stuff back. The Perfect example would be only your own files having changed.

Kane610 commented on homeassistant/components/ps4/config_flow.py:
```diff
@@ -0,0 +1,124 @@
+        if self.hass.config_entries.async_entries(DOMAIN):
+            return self.async_abort(reason='devices_configured')
+
+        # Check if able to bind to ports: UDP 987, TCP 997.
```
Is this for the app or would this be needed in after a restart with a configured entry as well? It is not clear why you bind the ports, improve documentation?

ktnrg45 replied:
I bind the ports just to check if the Home Assistant instance. If you cannot bind to the ports then there is no point in continuing. The most likely reason this would fail is if the python runtime does not have access to privileged ports and would need extra configuration described in my docs PR.

Kane610 commented on .coveragerc:
```diff
@@ -363,48 +279,22 @@ omit =
-    homeassistant/components/raspyrfm/*
+    homeassistant/components/nuimo_controller.py
+    homeassistant/components/prometheus.py
+    homeassistant/components/ps4/*
```
Your tests for config flow won't runt like this.

simonk83 commented:
Hey @ktnrg45, just wanted to say thanks for all your hard work with this, you're doing a great job 👍🏻

MartinHjelmare commented on .coveragerc:
```diff
@@ -15,28 +15,18 @@ omit =
     homeassistant/components/air_quality/nilu.py
     homeassistant/components/air_quality/norway_air.py
     homeassistant/components/air_quality/opensensemap.py
+    homeassistant/components/air_quality/nilu.py
```
The access mode has been changed on `.coveragerc`. That should be reverted.

There are also non related changes here, probably due to not rebasing properly.

MartinHjelmare commented on homeassistant/components/ps4/config_flow.py:
```diff
@@ -0,0 +1,124 @@
+        # Check if able to bind to ports: UDP 987, TCP 997.
+        failed = await self.hass.async_add_executor_job(helper.port_bind)
+        if failed is not None:
+            reason = 'port_{}_bind_error'.format(failed)
```
We shouldn't couple the ps4 api and the home assistant api like this. They should be independent.

Please add some logic here that converts the failure returned from the ps4 helper to the reason that we use when aborting.

One way could be a dictionary that maps the failure string to the reason string. If that is constant, put it at the module level.

ktnrg45 replied:
I'm not sure that I follow what you're saying. Can you explain further what you mean by coupling? 

I just call a function in the library. If it fails it returns the port number as an int. This is only so that the port which failed is displayed in the abort message.

Here is the method from the library:
```
    def port_bind(self):
        """Try binding to ports."""
        import socket
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.settimeout(3.0)
        ports = [987, 997]
        for port in ports:
            try:
                sock.bind(('0.0.0.0', port))
                sock.close()
            except socket.error:
                sock.close()
                return port
            return
```

MartinHjelmare replied:
I'm commenting on line 43. That's where the problem is.

We shouldn't use the result from an external api directly in our internal api. Either of those apis might change and we would then break.

We need to define the valid connections between the external api result and the internal api abort reason. Use a dictionary for that mapping.

MartinHjelmare replied:
Btw, the external library function returns inconsistently. Always return explicitly a value if any return statement returns a value.

ktnrg45 replied:
So like this:
*UDP_PORT and TCP_PORT are globals
```
    async def async_step_user(self, user_input=None):
        """Handle a user config flow."""
        # Abort if device is configured.
        msg = {
            UDP_PORT: 'port_987_bind_error',
            TCP_PORT: 'port_997_bind_error'}

        ports = msg.keys()

        if self.hass.config_entries.async_entries(DOMAIN):
            return self.async_abort(reason='devices_configured')

        # Check if able to bind to ports: UDP 987, TCP 997.
        failed = await self.hass.async_add_executor_job(
            self.helper.port_bind, ports)
        if failed in ports:
            reason = msg[failed]
            return self.async_abort(reason=reason)
        return await self.async_step_creds(user_input)
```

MartinHjelmare replied:
Move msg to a constant at the module level.

MartinHjelmare commented on homeassistant/components/ps4/config_flow.py:
```diff
@@ -0,0 +1,124 @@
+    async def async_step_creds(self, user_input=None):
+        """Return PS4 credentials from 2nd Screen App."""
+        from pyps4_homeassistant import Helper
+        helper = Helper()
```
Since we're using this helper instance in many methods, we could save it as an instance attribute on the flow handler. Then we wouldn't need to import and instantiate it multiple times.

ktnrg45 replied:
Ok. I was trying to follow the imports must be in functions where it is used rule.

MartinHjelmare replied:
If we import the helper class in init and save the helper instance as an attribute, we're following the rule.

MartinHjelmare commented on homeassistant/components/ps4/config_flow.py:
```diff
@@ -0,0 +1,124 @@
+
+        for device in devices:
+            ip_addr = device['host-ip']
+            device_list.append(ip_addr)
```
```py
device_list = [
    device['host-ip'] for device in devices]
```

MartinHjelmare commented on requirements_all.txt:
```diff
@@ -1,1814 +0,0 @@
-# Home Assistant core
```
Please revert this change. We should not delete this file.

MartinHjelmare commented:
There are merge conflicts.

ktnrg45 committed:
Added PS4/ __init__.py

ktnrg45 committed:
Added PS4/ __init__.py

ktnrg45 commented:
FYI did Checkout to before I made any commits and rebased.

ktnrg45 committed:
Create en.json

ktnrg45 committed:
Create en.json

ktnrg45 committed:
Create config_flow.py

ktnrg45 committed:
Create config_flow.py

ktnrg45 committed:
Create const.py

ktnrg45 committed:
Create const.py

ktnrg45 committed:
Create media_player.py

ktnrg45 committed:
Create media_player.py

ktnrg45 committed:
Create services.yaml

ktnrg45 committed:
Create services.yaml

ktnrg45 committed:
Create strings.json

ktnrg45 committed:
Create strings.json

ktnrg45 committed:
Create __init__.py

ktnrg45 committed:
Create __init__.py

ktnrg45 committed:
Add test_config_flow.py/ Finished adding PS4 files

ktnrg45 committed:
Add test_config_flow.py/ Finished adding PS4 files

ktnrg45 committed:
Rewrote for loop into short-hand

ktnrg45 committed:
Rewrote for loop into short-hand

ktnrg45 committed:
bumped pyps4 to 0.2.8

ktnrg45 committed:
bumped pyps4 to 0.2.8

ktnrg45 committed:
Pass in helper()

ktnrg45 committed:
Pass in helper()

MartinHjelmare commented on homeassistant/components/ps4/config_flow.py:
```diff
@@ -0,0 +1,116 @@
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self, helper):
```
This method shouldn't take arguments. We're not in control of the instantiation of this class.

ktnrg45 committed:
Rewrite func

ktnrg45 committed:
Rewrite func

victorcerutti committed:
Météo-France platform for the weather component (#18404)

* new weather component for meteofrance

* linting

* upgrade meteofrance package version

* Update .coveragerc

* Remove updates to the weather component architecture

* Rewrite Météo-France as a component

* Update .coveragerc

* Update requirements_all.txt

* remove Weather Card option

* Update conf name

Changing conf name to something more universal for worldwide weather forecast (postal code was only relevent for France)

* Update meteofrance pypi package

* fix line too long

* Update requirements_all.txt

* prevent from calling an API endpoint if not in monitored conditions

* fix stale url and remove blank line

* Insure that all cities are unique

* rename CONF_ATTRIBUTION

* Updating data from component setup

* fix missing extra lines

fabaff committed:
Update file header (#21061)

* Update file header

* Fix lint issue

* Fix lint issue

fabaff committed:
Update file header

ktnrg45 committed:
Fixed test

ktnrg45 committed:
Fixed test

ktnrg45 committed:
Added import in init

ktnrg45 committed:
Added import in init

ktnrg45 committed:
bump to 0.2.9

ktnrg45 committed:
bump to 0.2.9

ktnrg45 committed:
bump to 0.3.0

ktnrg45 committed:
bump to 0.3.0

ktnrg45 committed:
Removed line

ktnrg45 committed:
Removed line

ktnrg45 commented:
So I have to manually add the pypi library to gen_requirements_all.py in the list TEST_REQUIREMENTS right?

ktnrg45 committed:
lint

ktnrg45 committed:
lint

elupus committed:
Climate const.py move (#20945)

* Move constants to const.py

* Import from const instead of climate

MartinHjelmare commented:
Yes, but it's only needed if we need to import the library in the tests. We can also mock and patch the library. But it's ok to depend on the library and addit if we need for tests.

fabaff committed:
Update file header (#21067)

ktnrg45 committed:
Add ps4 to flows list

ktnrg45 committed:
Add ps4 to flows list

ktnrg45 committed:
Added pyps4-homeassistant with script

ktnrg45 committed:
Merge branch 'add-component-media_player.ps4' of https://github.com/ktnrg45/home-assistant into add-component-media_player.ps4
requirements_all.txt

ktnrg45 commented:
Reopening pr. Messed up rebase.
